src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -69,7 +72,33 @@ private FlagsHolder() {}
             GetBooleanAction.privilegedGetProperty("jdk.crac.trace-startup-time");
     }
 
-    private static final Context<Resource> globalContext = new BlockingOrderedContext<>();
+    private static final Context<Resource> globalContext = GlobalContext.createGlobalContextImpl();
+
+    private static class ReferenceHandlerResource implements JDKResource {
```

Comment:
I wouldn't pollute the `Core` class with domain-specific code; if the 'reliability' of registration is of concern I think it's fine to expose the static `ensureRegistered`/`register` method.
Or, at least refactor this into its own file.

Reply:
I considered register() interface. The waitForReferenceProcessing is an internal JDK interface. The reference processing itself does not require the special managment on checkpoint, but it's checkpoint requirement as a procedure that no reference processing is in progress at some certain point.

The class is a boilerplate code to call two methods, it is not shared and unlikely will, so a separate file is not justified.

Reply:
If it's a requirement (and not an optimization), how come it's not prevented from running? (the blocking was there for a time before you reverted that, wasn't it?)

Reply:
This piece of code is reletated to this part of the description

> To make properties available, it was required to delay Reference resource registration. Otherwise, the GlobalContext implementation decision had to be done too early, in the Reference initialization during JDK bootstrapping, before Properties are available.

Without this code, the ReferenceResource was registered during Reference initialization. At that point, Properties were not initalized yet, so any attempt to call System.getProperty() threw NPE (regardless parameters).

I.e. if we have the default context implementation configuration via properties, we need to delay ReferenceResource registration.

Reply:
I understand moving the registration from Reference clinit, and if you'd rather keep the implementation here, let's keep it that way. I was referring to

> it's checkpoint requirement as a procedure that no reference processing is in progress at some certain point

- I am not sure what creates this requirement, and it is not 100% guaranteed the way it's implemented now.

Reply:
The requirement is stated in the quoted part :) "no reference processing is in progress at some certain point" =(at REFERENCE_HANDLER priority).

It's required by the following CLEANERS (hard requirement, otherwise cleaner may miss something). And in general, reference processing should be quiescent -- to avoid checkpointing reference processing in the middle without a reason (soft requirement). I don't like how it's done at the moment. I've tried to make CLEANERS to ensure reference processing by its own, to be more straightforward. But in the absence of any CLEANER (e.g. the whole Cleaner package is not referenced and is not initialized), we won't satisfy the soft requirement, which still makes sense.